Client is not able to connect to cluster with advanced-network after split brain is healed

In case when split brain occurs and then is healed then client is not able to connect to cluster. This happens only when advanced-network configuration is used.

Following NPE is thrown on client side:
```
SEVERE: hz.client_0 [dev] [3.12-SNAPSHOT] Failed asynchronous execution of execution callback: com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback@6e488382for call ClientMessage{connection=null, length=174, correlationId=189, operation=Client.authentication, messageType=2, partitionId=-1, isComplete=true, isRetryable=true, isEvent=false, writeOffset=0}
java.lang.NullPointerException
    at java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936)
    at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$InetSocketAddressCache.get(ClientConnectionManagerImpl.java:788)
    at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$InetSocketAddressCache.access$200(ClientConnectionManagerImpl.java:782)
    at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$AuthCallback.onAuthenticated(ClientConnectionManagerImpl.java:733)
    at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$AuthCallback.onResponse(ClientConnectionManagerImpl.java:681)
    at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$AuthCallback.onResponse(ClientConnectionManagerImpl.java:643)
    at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:130)
    at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:118)
    at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:130)
    at com.hazelcast.client.spi.impl.ClientInvocationFuture$InternalDelegatingExecutionCallback.onResponse(ClientInvocationFuture.java:118)
    at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:256)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

Following exception on server side (on bigger part of merged cluster):
```
SEVERE: [192.168.86.29]:5702 [dev] [3.12-SNAPSHOT] hz._hzInstance_1_dev.event-3 caught an exception while processing:com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher@140d33b7
java.lang.IllegalArgumentException: argument 'address' can't be null
    at com.hazelcast.util.Preconditions.isNotNull(Preconditions.java:107)
    at com.hazelcast.instance.MemberImpl$Builder.<init>(MemberImpl.java:291)
    at com.hazelcast.client.impl.protocol.task.AddMembershipListenerMessageTask$MembershipListenerImpl.translateMemberAddress(AddMembershipListenerMessageTask.java:180)
    at com.hazelcast.client.impl.protocol.task.AddMembershipListenerMessageTask$MembershipListenerImpl.memberAdded(AddMembershipListenerMessageTask.java:122)
    at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.dispatchEvent(ClusterServiceImpl.java:781)
    at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.dispatchEvent(ClusterServiceImpl.java:85)
    at com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher.run(LocalEventDispatcher.java:64)
    at com.hazelcast.util.executor.StripedExecutor$Worker.process(StripedExecutor.java:244)
    at com.hazelcast.util.executor.StripedExecutor$Worker.run(StripedExecutor.java:227)
```

Following exception on server side (on smaller part of merged cluster):
```
INFO: [192.168.86.119]:5701 [dev] [3.12-SNAPSHOT] [192.168.86.119]:5701 is MERGED
Exception in thread "hz._hzInstance_1_dev.cached.thread-1" java.lang.NullPointerException: Target cannot be null!
    at com.hazelcast.util.Preconditions.checkNotNull(Preconditions.java:59)
    at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.createInvocationBuilder(OperationServiceImpl.java:279)
    at com.hazelcast.client.impl.ClientEngineImpl$DestroyEndpointTask.run(ClientEngineImpl.java:588)
    at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:227)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

There is also sometimes exception like that on client side (before aforementioned NPE):
```
INFO: hz.client_0 [dev] [3.12-SNAPSHOT] Authenticated with server [192.168.86.29]:9001, server version:3.12-SNAPSHOT Local address: /192.168.86.29:46725
Mar 26, 2019 9:19:32 AM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_0 [dev] [3.12-SNAPSHOT] Exception during initial connection to [192.168.86.29]:9001: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: argument 'address' can't be null
Mar 26, 2019 9:19:32 AM com.hazelcast.client.connection.nio.ClientConnection
WARNING: hz.client_0 [dev] [3.12-SNAPSHOT] ClientConnection{alive=false, connectionId=1, channel=NioChannel{/192.168.86.29:46725->/192.168.86.29:9001}, remoteEndpoint=[192.168.86.29]:9001, lastReadTime=2019-03-26 09:19:32.749, lastWriteTime=2019-03-26 09:19:32.746, closedTime=2019-03-26 09:19:32.751, connected server version=3.12-SNAPSHOT} closed. Reason: Could not connect to [192.168.86.29]:9001 as owner
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: argument 'address' can't be null
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:109)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:163)
	at com.hazelcast.client.spi.impl.ClientMembershipListener.listenMembershipEvents(ClientMembershipListener.java:143)
	at com.hazelcast.client.spi.impl.ClientClusterServiceImpl.listenMembershipEvents(ClientClusterServiceImpl.java:197)
	at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.onClusterConnect(HazelcastClientInstanceImpl.java:433)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.connectAsOwner(ClusterConnectorServiceImpl.java:150)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.connectToCandidate(ClusterConnectorServiceImpl.java:251)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.connectToClusterInternal(ClusterConnectorServiceImpl.java:195)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.access$000(ClusterConnectorServiceImpl.java:62)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl$1.call(ClusterConnectorServiceImpl.java:278)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl$1.call(ClusterConnectorServiceImpl.java:274)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Caused by: java.lang.IllegalArgumentException: argument 'address' can't be null
	at com.hazelcast.util.Preconditions.isNotNull(Preconditions.java:107)
	at com.hazelcast.instance.MemberImpl$Builder.<init>(MemberImpl.java:291)
	at com.hazelcast.client.impl.protocol.task.AddMembershipListenerMessageTask$MembershipListenerImpl.translateMemberAddress(AddMembershipListenerMessageTask.java:180)
	at com.hazelcast.client.impl.protocol.task.AddMembershipListenerMessageTask$MembershipListenerImpl.init(AddMembershipListenerMessageTask.java:107)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.addMembershipListener(ClusterServiceImpl.java:757)
	at com.hazelcast.client.impl.protocol.task.AddMembershipListenerMessageTask.call(AddMembershipListenerMessageTask.java:51)
	at com.hazelcast.client.impl.protocol.task.AbstractCallableMessageTask.processMessage(AbstractCallableMessageTask.java:35)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:124)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:104)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	... 17 more
```

**Steps to reproduce:**
1) Prepare member application - you need just `HazelcastInstance hz = Hazelcast.newHazelcastInstance();` and configuration file like:
```
<?xml version="1.0" encoding="UTF-8"?>
<hazelcast
        xsi:schemaLocation="http://www.hazelcast.com/schema/config http://www.hazelcast.com/schema/config/hazelcast-config-3.10.xsd"
        xmlns="http://www.hazelcast.com/schema/config"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

    <properties>
        <property name="hazelcast.merge.next.run.delay.seconds">30</property>
        <property name="hazelcast.merge.first.run.delay.seconds">50</property>
        <property name="hazelcast.operation.backup.timeout.millis">1000</property>
    </properties>
    <advanced-network enabled="true">
        <member-server-socket-endpoint-config>
            <interfaces enabled="true">
                <interface>SETUP_INTERFACE_HERE</interface>
            </interfaces>
            <port auto-increment="true">5701</port>
        </member-server-socket-endpoint-config>
        <client-server-socket-endpoint-config>
            <interfaces enabled="true">
                <interface>SETUP_INTERFACE_HERE</interface>
            </interfaces>
            <port auto-increment="true">9000</port>
        </client-server-socket-endpoint-config>
    </advanced-network>
</hazelcast>
```

2) Prepare client application - you need just `HazelcastInstance client = HazelcastClient.newHazelcastClient();` and configuration file like:
```
<?xml version="1.0" encoding="UTF-8"?>
<hazelcast-client
        xsi:schemaLocation="http://www.hazelcast.com/schema/client-config http://www.hazelcast.com/schema/client-config/hazelcast-client-config-3.10.xsd"
        xmlns="http://www.hazelcast.com/schema/client-config"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <network>
        <cluster-members>
            <address>SETUP_MEMBER_ADDRESS_HERE</address>
            <address>SETUP_MEMBER_ADDRESS_HERE</address>
            <address>SETUP_MEMBER_ADDRESS_HERE</address>
        </cluster-members>
    </network>
</hazelcast-client>
```

3) Start cluster with 2 members on computerA and 1 member on computerB. Start client on computerA.
4) Disconnect computerB from network and wait than split brain is detected.
5) Connect computerB back to network and wait than members form a cluster.
6) See that cluster state is not updated on client. Restart client - you will see aforementioned NPE and client is not connected to cluster.

--- a/hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java
@@ -582,7 +582,7 @@ public class ClientEngineImpl implements ClientEngine, CoreService, PreJoinAware
         @Override
         public void run() {
             InternalOperationService service = nodeEngine.getOperationService();
-            Address thisAddr = getThisAddress();
+            Address thisAddr = node.getThisAddress();
             for (Map.Entry<String, String> entry : ownershipMappings.entrySet()) {
                 String clientUuid = entry.getKey();
                 String memberUuid = entry.getValue();

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
@@ -26,6 +26,7 @@ import com.hazelcast.core.MemberSelector;
 import com.hazelcast.core.MembershipEvent;
 import com.hazelcast.core.MembershipListener;
 import com.hazelcast.hotrestart.HotRestartService;
+import com.hazelcast.instance.EndpointQualifier;
 import com.hazelcast.instance.HazelcastInstanceImpl;
 import com.hazelcast.instance.LifecycleServiceImpl;
 import com.hazelcast.instance.MemberImpl;
@@ -68,6 +69,7 @@ import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
@@ -319,11 +321,12 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
         assert !isJoined() : "Cannot reset local member UUID when joined.";
 
         Address memberAddress = node.getThisAddress();
+        Map<EndpointQualifier, Address> addressMap = localMember.getAddressMap();
         String newUuid = UuidUtil.createMemberUuid(memberAddress);
 
         logger.warning("Resetting local member UUID. Previous: " + localMember.getUuid() + ", new: " + newUuid);
 
-        MemberImpl memberWithNewUuid = new MemberImpl.Builder(memberAddress)
+        MemberImpl memberWithNewUuid = new MemberImpl.Builder(addressMap)
                 .version(localMember.getVersion())
                 .localMember(true)
                 .uuid(newUuid)

